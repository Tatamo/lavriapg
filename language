EXCLAMATION	"!"
VBAR		"|"
DOLLAR		"$"
COLON		":"
SEMICOLON	";"
LABEL		/[a-zA-Z_][a-zA-Z0-9_]*/
REGEXP		/\/.*\/[gimuy]*/
STRING		/".*"/
STRING		/'.*'/
!ENDLINE	/(\r\n|\r|\n)+/
!WHITESPACE	/[ \f\t\v\u00a0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000\ufeff]+/
INVALID	/./


$LANGUAGE :	LEX GRAMMAR;

LEX :		LEX LEXSECT | LEXSECT;
LEXSECT :	LEXLABEL LEXDEF;
LEXLABEL :	LABEL | EXCLAMATION | EXCLAMATION LABEL;
LEXDEF :	STRING | REGEXP;

GRAMMAR :	SECT GRAMMAR | SECT;
SECT :		SECTLABEL COLON DEF SEMICOLON;
SECTLABEL :	LABEL | DOLLAR LABEL;
DEF :		PATTERN VBAR DEF | PATTERN;
PATTERN :	SYMBOLLIST | ;
SYMBOLLIST :	LABEL SYMBOLLIST | LABEL;
