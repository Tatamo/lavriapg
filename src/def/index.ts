export * from "./language";
export * from "./parsingtable";
export * from "./token";
