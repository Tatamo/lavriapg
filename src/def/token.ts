/**
 * トークン名
 */
export type Token = string | symbol;

/**
 * トークン化された入力
 * トークン名と、字句規則にマッチした元々の入力
 */
export type TokenizedInput = { token: Token, value: string };

/**
 * 入力の終端を表す終端記号名
 * @type {symbol}
 */
export const SYMBOL_EOF: Token = Symbol("EOF");

/**
 * `S' -> S $` (Sは開始記号)となるような非終端記号S'を表す非終端記号名
 * @type {symbol}
 */
export const SYMBOL_SYNTAX: Token = Symbol("S'");
