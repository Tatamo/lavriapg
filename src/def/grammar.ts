import {Token} from "./token";
import {ILexer} from "../lexer/lexer";

export interface LexDefinitionSection {
	token: Token | null;
	pattern: string | RegExp;
	priority?: number;
	callback?: (lexer: ILexer, token: string | null, value: string, index: number) => any;
}

export type LexDefinitions = Array<LexDefinitionSection>;

export interface SyntaxDefinitionSection {
	ltoken: Token;
	pattern: Array<Token>;
}

export type SyntaxDefinitions = Array<SyntaxDefinitionSection>;

export interface GrammarDefinition {
	lex: LexDefinitions;
	syntax: SyntaxDefinitions;
	start_symbol: Token;
}
