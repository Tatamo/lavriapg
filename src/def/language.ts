import {Token} from "./token";
import {ILexer, LexController} from "../lexer/lexer";

/**
 * 字句解析器の状態を区別するためのラベル型
 */
export type LexStateLabel = string;

/**
 * デフォルトの字句解析器の状態
 */
export const default_lex_state = "default";

/**
 * 字句解析器に与える状態
 */
export interface LexState {
	label: LexStateLabel;
	is_exclusive?: boolean;
}

/**
 * 字句規則マッチ時に呼び出されるコールバック
 */
export type LexCallback = (token: string | null, value: string, lex: LexController) => [string | null, any] | { token: string | null, value: any } | string | null;

/**
 * 単一の字句ルール
 */
// TODO: tokenはlabelに名称変更してもよい？
export interface LexRule {
	token: Token | null;
	pattern: string | RegExp;
	state?: Array<LexStateLabel>;
	is_disabled?: boolean;
	priority?: number;
	callback?: LexCallback;
}

/**
 * 字句規則
 */
export interface LexDefinition {
	rules: Array<LexRule>;
	states?: Array<LexState>;
	default_callback?: LexCallback;
}

/**
 * 構文のreduce時に呼び出されるコールバック
 */
export type GrammarCallback = (children: Array<any>, token: string, lexer: ILexer) => any;

/**
 * 単一の構文ルール
 */
export interface GrammarRule {
	ltoken: Token;
	pattern: Array<Token>;
	callback?: GrammarCallback;
}

/**
 * 構文規則
 */
export interface GrammarDefinition {
	rules: Array<GrammarRule>;
	start_symbol: Token;
	default_callback?: GrammarCallback;
}

/**
 * 言語定義
 */
export interface Language {
	lex: LexDefinition;
	grammar: GrammarDefinition;
}
