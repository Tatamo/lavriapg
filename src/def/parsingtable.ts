import {Token} from "./token";

/**
 * Shiftオペレーション
 */
export type ShiftOperation = { type: "shift", to: number };

/**
 * Reduceオペレーション
 */
export type ReduceOperation = { type: "reduce", grammar_id: number };

/**
 * Shift/Reduceコンフリクト
 */
export type ConflictedOperation = { type: "conflict", shift_to: Array<number>, reduce_grammar: Array<number> };

/**
 * Acceptオペレーション
 */
export type AcceptOperation = { type: "accept" };

/**
 * Gotoオペレーション
 */
export type GotoOperation = { type: "goto", to: number };

/**
 * 構文解析器の実行する命令群
 */
export type ParsingOperation = ShiftOperation | ReduceOperation | ConflictedOperation | AcceptOperation | GotoOperation;

/**
 * 構文解析表
 */
export type ParsingTable = Array<Map<Token, ParsingOperation>>;
