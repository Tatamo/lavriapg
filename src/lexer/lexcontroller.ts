import {DEFAULT_LEX_STATE, Language, LexDefinition, LexRule, LexState, LexStateLabel} from "../def/language";

export type LexRuleLabel = string;

class LexRuleManager {
	// private states: { states: Map<LexStateLabel, LexState>, index: Map<LexStateLabel, Set<number>>, inheritance: Map<LexStateLabel, LexStateLabel> };
	private states: Map<LexStateLabel, { state: LexState, index: Set<number>, inheritance: LexStateLabel | null }>;
	private rules: { rules: Array<LexRule | undefined>, labels: Map<LexRuleLabel, number> };
	// 各ルールに一意なidを割り当てるためのカウンタ
	private id_counter: number;
	// ルールの削除によって割り当てがなくなったid
	private free_ids: Array<number>;
	constructor(language: Language) {
		const lex = language.lex;
		this.id_counter = 0;
		this.free_ids = [];

		// initialize lex states map
		this.states = new Map();
		// もしlexの定義内にデフォルト状態の記述があっても上書きされるだけなので問題ない
		this.addState({label: DEFAULT_LEX_STATE, is_exclusive: false});
		if (lex.states !== undefined) {
			for (const state of lex.states) {
				this.addState(state);
			}
		}

		// initialize lex rules
		this.rules = {rules: [], labels: new Map()};

		for (const rule of lex.rules.map((r) => LexRuleManager.formatLexRule(r))) {
			this.rules.rules[this.id_counter] = rule;
			// 状態ごとにインデックスを張る
			for (const state of rule.state!) {
				// TODO: statesに入れる
				if (!this.states.has(state)) {
					this.addState(state);
				}
				this.states.get(state)!.index.add(this.id_counter);
			}
			this.id_counter += 1;
		}
	}
	addState(label: LexStateLabel): boolean;
	addState(state: LexState): boolean;
	addState(s: LexStateLabel | LexState): boolean {
		let state: LexState;
		if (typeof s === "object") {
			state = s;
		}
		else {
			state = {label: s};
		}
		state = LexRuleManager.formatLexState(state);
		if (this.states.has(state.label)) {
			return false;
		}
		const inheritance: LexStateLabel | null = state.is_exclusive ? null : DEFAULT_LEX_STATE;
		this.states.set(state.label, {state, index: new Set(), inheritance});
		// ループチェック
		if (inheritance !== null) {
			let flg_loop = false;
			let parent = this.states.get(inheritance);
			while (parent !== undefined && parent.inheritance !== null) {
				// 状態を追加するたびにチェックするので、自身にたどりつかないことを調べればよい
				if (parent.inheritance === state.label) {
					flg_loop = true;
					break;
				}
				parent = this.states.get(parent.inheritance);
			}
			if (flg_loop) this.states.get(state.label)!.inheritance = null;
		}
		return true;
	}
	// TODO: パフォーマンス改善
	getRulesItr(state: LexStateLabel): IterableIterator<LexRule> {
		// そんな状態はない
		if (!this.states.has(state)) return [][Symbol.iterator]();

		// 継承を加味
		let result: Array<number> = [];
		let s = this.states.get(state);
		while (s !== undefined) {
			result = result.concat([...s.index]);
			if (s.inheritance === null) break;
			s = this.states.get(s.inheritance);
		}

		return (function* (self, itr) {
			for (const id of itr) {
				if (self.rules.rules[id] !== undefined) yield self.rules.rules[id]!;
			}
		})(this, new Set(result)[Symbol.iterator]());
	}
	addRule(label: LexRuleLabel, rule: LexRule): void {
		// 同名の既存ルールを破棄
		this.removeRule(label);

		const formatted_rule = LexRuleManager.formatLexRule(rule);

		const id = this.free_ids.length > 0 ? this.free_ids.pop()! : this.id_counter++;
		this.rules.rules[id] = formatted_rule;
		this.rules.labels.set(label, id);
		for (const state of formatted_rule.state!) {
			if (!this.states.has(state)) this.addState(state);
			this.states.get(state)!.index.add(id);
		}
	}
	removeRule(label: LexRuleLabel): LexRule | null {
		if (!this.rules.labels.has(label)) {
			return null;
		}
		const id = this.rules.labels.get(label)!;
		const rule = this.rules.rules[id];
		if (rule === undefined) return null;

		for (const state of rule.state!) {
			if (this.states.has(state)) {
				this.states.get(state)!.index.delete(id);
			}
		}
		this.rules.rules[id] = undefined;
		this.free_ids.push(id);
		return rule;
	}
	static formatLexState(state: LexState): LexState {
		return {
			label: state.label,
			is_exclusive: state.is_exclusive !== undefined ? state.is_exclusive : false
		};
	}
	static formatLexRule(rule: LexRule): LexRule {
		// clone rule
		const result = {...rule};
		if (result.is_disabled === undefined) result.is_disabled = false;
		// 状態指定を省略された場合はデフォルト状態のみとする
		if (result.state === undefined) result.state = [DEFAULT_LEX_STATE];
		// 正規表現を字句解析に適した形に整形
		if (result.pattern instanceof RegExp) {
			result.pattern = LexRuleManager.formatRegExp(result.pattern);
		}
		return result;
	}
	private static formatRegExp(pattern: RegExp): RegExp {
		// フラグを整形する
		let flags: string = "";
		// gフラグは邪魔なので取り除く
		// i,m,uフラグがあれば維持する
		if (pattern.ignoreCase) {
			flags += "i";
		}
		if (pattern.multiline) {
			flags += "m";
		}
		if (pattern.unicode) {
			flags += "u";
		}
		// yフラグは必ずつける
		flags += "y";
		return new RegExp(pattern, flags);
	}
}

export class LexController {
	private _lex: LexDefinition;
	private _current_state: LexStateLabel;
	private _state_stack: Array<LexStateLabel>;
	private _rules: LexRuleManager;
	constructor(language: Language) {
		this._lex = language.lex;
		this._current_state = DEFAULT_LEX_STATE;
		this._state_stack = [];
		this._rules = new LexRuleManager(language);
	}
	onBegin(): void {
		if (this._lex.begin_callback !== undefined) this._lex.begin_callback(this);
	}
	onEnd(): void {
		if (this._lex.end_callback !== undefined) this._lex.end_callback(this);
	}
	getRulesItr(): IterableIterator<LexRule> {
		return this._rules.getRulesItr(this._current_state);
	}
	addRule(label: string, rule: LexRule): void {
		this._rules.addRule(label, rule);
	}
	removeRule(label: string): LexRule | null {
		return this._rules.removeRule(label);
	}
	getCurrentState(): LexStateLabel {
		return this._current_state;
	}
	jumpState(state: LexStateLabel): void {
		this._current_state = state;
	}
	callState(state: LexStateLabel): void {
		this._state_stack.push(this._current_state);
		this._current_state = state;
	}
	returnState(): LexStateLabel | undefined {
		const pop = this._state_stack.pop();
		if (pop === undefined) this._current_state = DEFAULT_LEX_STATE;
		else this._current_state = pop;
		return pop;
	}
}
