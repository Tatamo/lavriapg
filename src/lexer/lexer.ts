import {LexCallback, LexDefinition, LexRule} from "../def/language";
import {SYMBOL_EOF, Token, TokenizedInput} from "../def/token";
import {CallbackController} from "../parser/callback";

// TODO: 要改善
export interface ILexer {
	exec(str: string): Array<TokenizedInput>;
	setCallbackController(cc: CallbackController): void;
}

export class Lexer implements ILexer {
	private _rule_id: number;
	private _def: Array<{ id: number, rule: LexRule }>;
	private _reset: { def: Array<{ id: number, rule: LexRule }>, id: number, flg_modified: boolean };
	private _input: string;
	private _last_index: number;
	get last_index(): number {
		return this._last_index;
	}
	private _status: "uninitialized" | "ready" | "finished";
	get status(): "uninitialized" | "ready" | "finished" {
		return this._status;
	}
	private callback_controller: CallbackController;
	constructor(def: LexDefinition, input?: string) {
		this._def = [];
		this._rule_id = 0;
		for (const rule of def) {
			this.add(rule);
		}
		this.setCurrentDefinitionAsDefault();
		this.reset(input);
	}
	public setCallbackController(cc: CallbackController) {
		this.callback_controller = cc;
	}
	// リセット時に現在の字句規則になるようにする
	setCurrentDefinitionAsDefault() {
		this._reset = {
			def: this._def.slice(),
			id: this._rule_id,
			flg_modified: false
		};
	}
	// Lexerの内部状態をコンストラクタ呼び出し直後まで戻す
	reset(input?: string) {
		if (this._reset.flg_modified) {
			this._def = this._reset.def.slice();
			this._rule_id = this._reset.id;
			this._reset.flg_modified = false;
		}
		this._last_index = 0;
		this._status = "uninitialized";
		if (input !== undefined) this.init(input);
	}
	// 入力を与えて解析可能(step()を呼び出し可能)な状態にする
	init(input: string) {
		this._input = input;
		this._last_index = 0;
		this._status = "ready";
	}
	// 入力からトークン1つ分読み込む
	step(): TokenizedInput {
		if (this.status !== "ready") {
			throw new Error("Lexer is not ready");
		}
		while (true) {
			if (this._last_index >= this._input.length) {
				// 解析終了
				this._last_index = this._input.length;
				this._status = "finished";
				// 最後にEOFトークンを付与
				return {token: SYMBOL_EOF, value: ""};
			}
			let flg_matched = false;
			let result_token: Token | null = null;
			let result_match: string = "";
			let result_priority: number | null = null;
			let next_index: number;
			let result_callback: LexCallback | undefined;
			let result_id: number = -1;
			for (const {id, rule} of this._def) {
				const {token, pattern, priority, callback} = rule;
				let match: string;
				let tmp_next_index: number;
				if (typeof pattern === "string") {
					match = pattern;
					tmp_next_index = this._last_index + pattern.length;
					if (this._input.substring(this._last_index, tmp_next_index) != pattern) continue; // マッチしない
					if (tmp_next_index < this._input.length && /\w/.test(pattern.slice(-1)) && /\w/.test(this._input[tmp_next_index])) continue; // マッチした文字列の末尾が\wで、その直後の文字が\wの場合はスキップ
					flg_matched = true;
				}
				else {
					// pattern: RegExp
					pattern.lastIndex = this._last_index;
					const m = pattern.exec(this._input);
					if (m === null) continue; // マッチ失敗
					flg_matched = true;
					match = m[0];
					tmp_next_index = pattern.lastIndex;
				}
				// 同じ優先度の場合、最長マッチまたは出現順(match_priorityで設定)
				const _priority = priority !== undefined ? priority : 0;
				if (result_priority === null ||
					_priority > result_priority ||
					_priority == result_priority && match.length > result_match.length) {
					result_token = token;
					result_match = match;
					result_priority = _priority;
					next_index = tmp_next_index;
					result_callback = callback;
					result_id = id;
				}
			}
			if (flg_matched) {
				// 読む位置を進める
				this._last_index = next_index!;
				// コールバック呼び出し
				if (typeof result_token !== "symbol") {
					if (this.callback_controller !== undefined) {
						if (result_token !== null) {
							return {token: result_token, value: this.callback_controller.callLex(result_id, result_match, this)};
						}
						else {
							this.callback_controller.callLex(result_id, result_match, this);
						}
					}
					// CallbackControllerが定義されていないならそのままコールバックを呼び出す(暫定)
					else if (result_callback !== undefined) {
						if (result_token !== null) {
							return {token: result_token, value: result_callback(result_match, result_token, this)};
						}
						else {
							result_callback(result_match, result_token, this);
						}
					}
				}
				// tokenがnullなら処理を飛ばす
				if (result_token !== null) {
					return {token: result_token, value: result_match};
				}
			}
			else {
				// マッチする規則がなかった
				throw new Error("no pattern matched");
			}
		}
	}
	exec(input?: string): Array<TokenizedInput> {
		if (this.status === "finished" || this.status === "ready" && this.last_index > 0) {
			this.reset(input);
		}
		else if (input !== undefined) {
			this.init(input);
		}
		const result: Array<TokenizedInput> = [];
		while (this.status !== "finished") {
			result.push(this.step());
		}
		return result;
	}
	// 字句規則を追加し、そのidを返す
	add(rule: LexRule): number {
		const id = this._rule_id++;
		const token_pattern = rule.pattern;
		// 正しいトークン定義が与えられているかチェック
		if (typeof token_pattern === "string") {
			this._def.push({id, rule});
		}
		else if (token_pattern instanceof RegExp) {
			// フラグを整形する
			let flags: string = "";
			// gフラグは邪魔なので取り除く
			// i,m,uフラグがあれば維持する
			if (token_pattern.ignoreCase) {
				flags += "i";
			}
			if (token_pattern.multiline) {
				flags += "m";
			}
			if (token_pattern.unicode) {
				flags += "u";
			}
			// yフラグは必ずつける
			flags += "y";
			this._def.push({
				id: id,
				rule: {
					token: rule.token,
					pattern: new RegExp(token_pattern, flags),
					priority: rule.priority,
					callback: rule.callback
				}
			});
		}
		else {
			throw new Error("invalid token definition: neither string nor RegExp object");
		}
		if (this._reset !== undefined) this._reset.flg_modified = true;
		return id;
	}
	// 字句規則を削除する
	// TODO: もっとましな実装にする
	del(id: number): LexRule {
		for (let i = 0; i < this._def.length; i++) {
			if (this._def[i].id === id) {
				this._reset.flg_modified = true;
				return this._def.splice(i, 1)[0].rule;
			}
		}
		throw new Error("definition not found");
	}
}
