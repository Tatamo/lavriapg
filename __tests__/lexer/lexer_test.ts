import {Lexer} from "../../src/lexer/lexer";
import {test_empty_language, test_sample_language} from "../data/sample_language";
import {SYMBOL_EOF} from "../../src/def/token";
import {Language} from "../../src/def/language";

describe("Lexer test", () => {
	const empty_lang: Language = {lex: {rules: []}, grammar: {rules: [], start_symbol: ""}};
	test("exec valid input", () => {
		const lexer = new Lexer(test_sample_language);
		expect(lexer.exec("xabc;x|&0ax x z;")).toMatchSnapshot();
	});
	test("exec invalid input", () => {
		const lexer = new Lexer(test_empty_language);
		expect(() => {
			lexer.exec("xabc;x|&0ax x z;");
		}).toThrow(/no pattern matched/);
	});
	test("exec no length input", () => {
		const lexer = new Lexer(test_sample_language);
		expect(lexer.exec("")).toEqual([
			{token: SYMBOL_EOF, value: ""}
		]);
		const lexer2 = new Lexer(test_empty_language);
		expect(lexer2.exec("")).toEqual([
			{token: SYMBOL_EOF, value: ""}
		]);
	});
	test("regexp flags", () => {
		const lexer = new Lexer({
			grammar: {rules: [], start_symbol: ""}, lex: {
				rules: [
					{token: "I", pattern: /AbC/i},
					{token: "M", pattern: /x\nyz/m},
					{token: "U", pattern: /\u{64}\u{65}\u{66}/u},
					{token: "G", pattern: /pqr/g},
					{token: "A", pattern: /\u{61}\nC/imugy}
				]
			}
		});
		expect(lexer.exec("abcx\nyzdefpqra\nc")).toMatchSnapshot();
	});
	test("skip string pattern if the following is \\w", () => {
		const lexer = new Lexer({
			grammar: {rules: [], start_symbol: ""}, lex: {
				rules: [
					{token: "STR", pattern: "abc"},
					{token: "REGEXP", pattern: /abc/},
					{token: "ASTERISK", pattern: "*"},
					{token: "XYZ", pattern: "xyz"}
				]
			}
		});
		expect(lexer.exec("abcxyz*abc*xyz*abcabc")).toMatchSnapshot();
	});
	test("rule priority", () => {
		const lexer = new Lexer({
			grammar: {rules: [], start_symbol: ""}, lex: {
				rules: [
					{token: "PM", pattern: "+-"},
					{token: "PMA", pattern: "+-*"},
					{token: "ASTERISK", pattern: "*", priority: 1},
					{token: "ABC", pattern: /abc/},
					{token: "ABCD", pattern: /abcd/},
					{token: "ABCD2", pattern: /abcd/, priority: 2},
					{token: "D", pattern: /d/},
					{token: "XYZ", pattern: /xyz/},
					{token: "XYZW", pattern: /xyzw/, priority: -1},
					{token: "W", pattern: /w/},
					{token: null, pattern: " "}
				]
			}
		});
		expect(lexer.exec(" +-+-*abcd xyzw")).toMatchSnapshot();
	});
	test("longest match", () => {
		const lexer = new Lexer({
			grammar: {rules: [], start_symbol: ""}, lex: {
				rules: [
					{token: "PM", pattern: "+-"},
					{token: "PMA", pattern: "+-*"},
					{token: "ASTERISK", pattern: "*"},
					{token: "ABC", pattern: /abc/},
					{token: "ABCD", pattern: /abcd/},
					{token: "ABCD2", pattern: /abcd/},
					{token: "D", pattern: /d/},
					{token: null, pattern: " "}
				]
			}
		});
		expect(lexer.exec(" +-+-*abcd ")).toMatchSnapshot();
	});
	test("callbacks", () => {
		// 引数として与えられるLexControllerを使用した詳細なテストはcontroller_test.tsで
		const lexer = new Lexer({
			grammar: {rules: [], start_symbol: ""}, lex: {
				rules: [
					{token: "A", pattern: /a/},
					{token: "B", pattern: /b/, callback: (value, token) => token},
					{token: "C", pattern: /c/, callback: (value, token) => ({token, value: "2"})},
					{
						token: "D", pattern: /d/,
						callback: () => {
							return;
						}
					},
					{token: "E", pattern: /e/, callback: () => null},
					{token: null, pattern: " "}
				],
				default_callback: (value, token) => {
					return [token, "1"];
				}
			}
		});
		expect(lexer.exec("abc de")).toEqual([
			{token: "A", value: "1"},
			{token: "B", value: "b"},
			{token: "C", value: "2"},
			{token: "D", value: "d"},
			{token: SYMBOL_EOF, value: ""}
		]);
	});
});
