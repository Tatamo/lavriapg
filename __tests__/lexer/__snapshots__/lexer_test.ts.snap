// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lexer test exec valid input 1`] = `
Array [
  Object {
    "token": "ID",
    "value": "xabc",
  },
  Object {
    "token": "SEMICOLON",
    "value": ";",
  },
  Object {
    "token": "ATOM",
    "value": "x",
  },
  Object {
    "token": "SEPARATE",
    "value": "|",
  },
  Object {
    "token": "INVALID",
    "value": "&",
  },
  Object {
    "token": "INVALID",
    "value": "0",
  },
  Object {
    "token": "ID",
    "value": "ax",
  },
  Object {
    "token": "ATOM",
    "value": "x",
  },
  Object {
    "token": "ID",
    "value": "z",
  },
  Object {
    "token": "SEMICOLON",
    "value": ";",
  },
  Object {
    "token": Symbol(EOF),
    "value": "",
  },
]
`;

exports[`Lexer test longest match 1`] = `
Array [
  Object {
    "token": "PM",
    "value": "+-",
  },
  Object {
    "token": "PMA",
    "value": "+-*",
  },
  Object {
    "token": "ABCD",
    "value": "abcd",
  },
  Object {
    "token": Symbol(EOF),
    "value": "",
  },
]
`;

exports[`Lexer test regexp flags 1`] = `
Array [
  Object {
    "token": "I",
    "value": "abc",
  },
  Object {
    "token": "M",
    "value": "x
yz",
  },
  Object {
    "token": "U",
    "value": "def",
  },
  Object {
    "token": "G",
    "value": "pqr",
  },
  Object {
    "token": "A",
    "value": "a
c",
  },
  Object {
    "token": Symbol(EOF),
    "value": "",
  },
]
`;

exports[`Lexer test rule priority 1`] = `
Array [
  Object {
    "token": "PM",
    "value": "+-",
  },
  Object {
    "token": "PMA",
    "value": "+-*",
  },
  Object {
    "token": "ABCD2",
    "value": "abcd",
  },
  Object {
    "token": "XYZ",
    "value": "xyz",
  },
  Object {
    "token": "W",
    "value": "w",
  },
  Object {
    "token": Symbol(EOF),
    "value": "",
  },
]
`;

exports[`Lexer test skip string pattern if the following is \\w 1`] = `
Array [
  Object {
    "token": "REGEXP",
    "value": "abc",
  },
  Object {
    "token": "XYZ",
    "value": "xyz",
  },
  Object {
    "token": "ASTERISK",
    "value": "*",
  },
  Object {
    "token": "STR",
    "value": "abc",
  },
  Object {
    "token": "ASTERISK",
    "value": "*",
  },
  Object {
    "token": "XYZ",
    "value": "xyz",
  },
  Object {
    "token": "ASTERISK",
    "value": "*",
  },
  Object {
    "token": "REGEXP",
    "value": "abc",
  },
  Object {
    "token": "STR",
    "value": "abc",
  },
  Object {
    "token": Symbol(EOF),
    "value": "",
  },
]
`;
